File: packages/api/src/type/ws.ts
import { z } from "zod/v4";

//i don't know what the others is
export const WSMessageType = ["music", "ping", "others"] as const;

export const WSMusicAction = ["play", "pause", "change", "changeTime"] as const;

export const WSMusicMessageClientPayloadSchema = z.object({
  action: z.enum(WSMusicAction),
  positionMs: z.number().min(0),
  trackID: z.string().optional(),
});

export const WSMusicMessageServerPayloadSchema = z.object({
  action: z.enum(WSMusicAction),
  userID: z.string(),
  positionMs: z.number().min(0),
  trackID: z.string(),
  ServerTime: z.number().min(0),
});

export type WSMusicMessageClientPayloadType = z.infer<
  typeof WSMusicMessageClientPayloadSchema
>;

File: desktop/packages/presence/src/types/ws.rs
use serde::Deserialize;

#[derive(Debug, Deserialize, PartialEq)]
pub enum WSMusicAction {
    #[serde(rename = "play")]
    Play,
    #[serde(rename = "pause")]
    Pause,
    #[serde(rename = "change")]
    Change,
    #[serde(rename = "changeTime")]
    ChangeTime,
}

#[derive(Clone, Debug, Deserialize)]
pub struct TrackData {
    #[serde(rename = "trackID")]
    pub track_id: String,
    pub cover: Option<String>,
    #[serde(rename = "albumName")]
    pub album_name: String,
    #[serde(rename = "trackName")]
    pub title: String,
    pub artists: Vec<String>,
    pub duration: u64,
    #[serde(rename = "noOfTracks")]
    pub no_of_tracks: u64,
    #[serde(rename = "trackNo")]
    pub track_no: u64,
}

#[derive(Debug, Deserialize)]
#[serde(tag = "type", content = "payload")]
pub enum WSMessage {
    #[serde(rename = "music")]
    Music(WSMusicMessageServerPayload),
    #[serde(rename = "others")]
    Others(),
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WSMusicMessageServerPayload {
    pub action: WSMusicAction,
    #[serde(rename = "userID")]
    pub _user_id: String,
    pub position_ms: f64,
    #[serde(rename = "trackID")]
    pub track_id: String,
    #[serde(rename = "ServerTime")]
    pub server_time: u64,
}
